### PR TITLE
Order history can track modifications via api calls

### LINKED ISSUE #34830: Adapt the command Order status for the API access
id_employee | order_history |

#### Specs

- create the endpoint:
/orders/{orderId}/update-status
/orders/bulk-update-status

- Adapt the command to be able to save API access and associated to the API ID
  - BulkChangeOrderStatusCommand
  - UpdateOrderStatusCommand

- Add the new column in table order_history to store the ID API access 

- Adapt the page to display the API access name if the status is updated by the API.

#### Acceptance 

When an API access updates the order status 
Then the API access name is displayed on the order page in the order status history next to the new order status

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Order history can track modifications via api calls
| Type?             | new feature
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| How to test?      | CI green, UI tests green Create an order, then create an ApiAccess to modify the order status thanks to the API, in the BO page the status modification should be assigned to an ApiClient
| UI Tests          | https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/7922318247
| Fixed issue or discussion?     | Fixes #34830
| Related PRs       | Module PR incoming to add the order endpoint
| Sponsor company   | ~

### Breaking changes
- `EmployeeId` now has a strict type `int` in `getValue` method
- `OrderStatusForViewing` constructor requires an additional parameter

### CODE REVIEW

**Review by Hlavtox — CHANGES_REQUESTED** [review]
This must be done by a centralized logging system for all entities, not adding a new database column. It doesn't belong there. Logs are usually cleaned in order to shrink database size.

**Comment by MatShir** [comment]
So we have mainly 3 commands to adapts to make sure the core will work properly for the API where an employee ID is expected:
- Adapt the command Order status for the API access  #34830
- Adapt StockManager for the API access #34832
- Adapt the command Order payment for the API access #34831

Is it worth blocking thoses 3 adpations now for a new log system? Maybe there is a commun ground that could be found there 🤷 

Also, so from your logic, it means that the order status history even from employee action on the order page needs to be transfered into the new log system as well ?

**Comment by Hlavtox** [comment]
@MatShir The legacy code allows to set order statuses with no employee ID - for example, like modules do it, with no specific employee information. There shouldn't be a need to specify employee ID for any action.

**Comment by MatShir** [comment]
Sad not to willing see who performed the action specially for order status and to take a look on the system log

**Comment by Hlavtox** [comment]
@MatShir I am willing to see it, I proposed this many years ago, that we must see what employee did what.

But it must be done properly, not randomly adding a some column with api ID into order state history, not consulting it with anyone, not even other SA employees.

A logging system can look like this - every action that the employeee did.
Snímek obrazovky 2024-02-08 114541

Or for a specific object - every change that was done on the entity.
Snímek obrazovky 2024-02-08 114651

**Comment by MatShir** [comment]
I always agreed with it, but improvement needs to be done step by step. And it won't be done in the minute, so you prefer having no id why there is a change on the order status rather than collecting the id of the app ?

**Review by kpodemski — CHANGES_REQUESTED** [review]
I don't see it as an improvement, sorry.

I'm clicking "Request changes", but feel free to reject it.

I already commented why I think it's a wrong idea in the Issue. I understand that we want to introduce improvements "step by step", but there's no easy way back from this.

People will see this and will try to make their entities with api_client_id because they'll think it's a "best practice".

For 15 years, there were no logs for webservices, you can have order history changed or stock changed without any indicator of who has done it. Now, with a new API, it is not needed - in my opinion - to come up with interim solution just because we need logs anyway.

**Comment by kpodemski** [comment]
>  And it won't be done in the minute, so you prefer having no id why there is a change on the order status rather than collecting the id of the app ?

I don't see how introducing a new tab, grid, database table, is much more time-consuming than what I see in this PR 🤔 

Creating "API -> Logs" page does not mean we need to come up with the perfect solution for logging every single action and set in stone. What's more, it could be introduced behind a feature flag or as a setting on whether to log operations or not, how often clean the logs, with a filtering that comes with a Grid etc., etc.

**Comment by jolelievre** [comment]
I think you're both confusing logs and attaching order history to the API Access 🤔

Seems like you're talking about generic logs for each action which I understand and that could be cleaned over time when they grow large which I also agree. But this is not what this PR is about it's about hard-linking a modification stored in the DB to the person that dit it, previously it was an employee now it needs to be an API client as well. And this shouldn't be cleaned IMHO as it's part of the order history, so it's not exactly the same thing as a log.

But I'd be happy to discuss more about this with you two, there's definitely room for improvement, we could store the association in a dedicated table/entity instead of adding a column everytime we need to link an action. So maybe we could even think about a generic way of assiging an action to an API client (and why not later to an employee), and maybe that's what you have in mind when you talk about logs, anyway it's worth discussing about it.

**Comment by kpodemski** [comment]
> But I'd be happy to discuss more about this with you two, there's definitely room for improvement, we could store the association in a dedicated table/entity instead of adding a column everytime we need to link an action. So maybe we could even think about a generic way of assiging an action to an API client (and why not later to an employee), and maybe that's what you have in mind when you talk about logs, anyway it's worth discussing about it.

Exactly that. This way, you are not polluting core tables with new columns, and it would also be - in my opinion - easier to work with.

**Comment by jolelievre** [comment]
OK @kpodemski I'll think of a more generic alternative then 😉

File: src/Core/Mutation/MutationTracker.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    public function addMutation(string $table, int $rowId, string $action, int $moduleId = null): void
+    {
+        $mutation = new Mutation();
+        $mutation
+            ->setTable($table)
+            ->setRowId($rowId)
+            ->setAction($action)
+            ->setEmployeeId($this->employeeContext->getEmployee()?->getId())
+            ->setApiClientId($this->apiClientContext->getApiClient()?->getId())
+            ->setModuleId($moduleId)
```
Comment (kpodemski) [inline src/Core/Mutation/MutationTracker.php]: I believe we should not track the id of the module since it can change. My understanding of this mechanism is that we should be able to show the historical data, meaning it should be the name of the module (possibly technical name).

File: src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php [hunk 2]
```diff
@@ -392,7 +392,8 @@ private function getOrderHistory(Order $order): OrderHistoryForViewing
                 new DateTimeImmutable($item['date_add']),
                 (bool) $item['send_email'],
                 $item['employee_firstname'],
-                $item['employee_lastname']
+                $item['employee_lastname'],
+                $item['api_client_id'],
```
Comment (kpodemski) [inline src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:396]: Could we be less explicit here and retrieve mutation details instead of `api_client_id`? Imagine if you want to show a box showing a given order's mutations list.

File: src/Core/Mutation/MutationTracker.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+     */
+    public function addMutation(string $table, int $rowId, string $action, int $moduleId = null): void
+    {
+        $mutation = new Mutation();
+        $mutation
+            ->setTable($table)
+            ->setRowId($rowId)
+            ->setAction($action)
+            ->setEmployeeId($this->employeeContext->getEmployee()?->getId())
+            ->setApiClientId($this->apiClientContext->getApiClient()?->getId())
```
Comment (kpodemski) [inline src/Core/Mutation/MutationTracker.php]: Food for thought: imagine that you use this MutationTracker not only to work with a new API but also with web services. What about taking it a step further and making it more universal?

```
->setActorId($actorContext->getActorId())
->setActorType($actorContext->getActorType())
```

and something like:
`private function getActorContext(): ActorContextInterface`

File: src/Core/Mutation/MutationTracker.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+     * @param string $table
+     * @param int $rowId
+     * @param string $action
+     * @param int|null $moduleId
+     */
+    public function addMutation(string $table, int $rowId, string $action, int $moduleId = null): void
+    {
+        $mutation = new Mutation();
+        $mutation
+            ->setTable($table)
```
Comment (kpodemski) [inline src/Core/Mutation/MutationTracker.php]: I think that for the sake of consistency, we should make sure we add or remove the prefix from the database.

File: src/Core/Mutation/MutationTracker.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ * but as long as these parameters are empty the default value will always be set if present, so if you
+ * need more accurate control on this association you should create and persist the mutation manually.
+ *
+ * It also contains a few action constants as the most usually used, but the action remains a customizable
+ * field.
+ */
+class MutationTracker
+{
+    public const CREATE_ACTION = 'create';
+    public const UPDATE_ACTION = 'update';
```
Comment (kpodemski) [inline src/Core/Mutation/MutationTracker.php]: This may be a controversial proposal, but do you think we could log a real difference between the old and new states of the given row? Handling it automatically could be PITA, but what about having an old and new state passed to the MutationTracker? 🤔 

A very simple thing I do in one of the projects:

```
$changes = [
    'fieldName1' => ['old' => $oldValue1, 'new' => $newValue1],
    'fieldName2' => ['old' => $oldValue2, 'new' => $newValue2],
];
```

so it could be:

`$mutation->setChanges(json_encode($changes));`

Of course, this should be opt-in as it may drastically increase the size of the table.

Edit after the meeting:
- since it is not about logging stuff it is not needed

File: src/Core/Mutation/MutationTracker.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+     * performing the mutation.
+     *
+     * @param string $table
+     * @param int $rowId
+     * @param string $action
+     * @param int|null $moduleId
+     */
+    public function addMutation(string $table, int $rowId, string $action, int $moduleId = null): void
+    {
+        $mutation = new Mutation();
```
Comment (kpodemski) [inline src/Core/Mutation/MutationTracker.php:110]: A few things to consider later:
- BO interface for the data

edited after the meeting

File: src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php (hunk 2 above)
Comment (kpodemski) [reply src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:396]: Or there should not be anything there and just a way to click and be redirected to show the details of a given mutation 🤔 I'm not sure; it depends on how you would like to show the details in the interface. Is there any plan to show explicit information that a given action was performed by an API client?

File: src/Core/Mutation/MutationTracker.php (hunk 6 above)
Comment (Hlavtox) [reply src/Core/Mutation/MutationTracker.php:110]: Yes, disabled by default, enabled opt-in, cleaning by cron required/recommended.

File: src/Core/Mutation/MutationTracker.php (hunk 5 above)
Comment (Hlavtox) [reply src/Core/Mutation/MutationTracker.php]: Also, we should support all types of actions - delete, change configuratuon, install module, remove module etc

Comment (kpodemski) [reply src/Core/Mutation/MutationTracker.php]: but it is not about module, there are only three actions but who performed them would be different

File: src/Core/Mutation/MutationTracker.php (hunk 6 above)
Comment (jolelievre) [reply src/Core/Mutation/MutationTracker.php:110]: I ans<er this for people reading the thread, but as we discussed this mutation relationship os not meant for automatic logs It really represents an relationship that has true meaning from the domain perspective, we display this history information in the BO, the purpose of this ne relation entity is not to be used for random tracking everywhere

It's used for meaningful interactions between some entities that need to be historized and displayed in som pages

File: src/PrestaShopBundle/Entity/Mutation.php [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+ * @ORM\Entity(repositoryClass="PrestaShopBundle\Entity\Repository\MutationRepository")
+ */
+class Mutation
+{
+    /**
+     * @ORM\Id
+     * @ORM\Column(name="id_mutation", type="integer", options={"unsigned": true})
+     * @ORM\GeneratedValue(strategy="AUTO")
+     */
+    private int $id;
```
Comment (kpodemski) [inline src/PrestaShopBundle/Entity/Mutation.php:56]:
```suggestion
```

I think we agreed not to use `ID` for this table. Is there any usage for it?

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/history.html.twig [hunk 8]
```diff
@@ -38,13 +38,16 @@
         {% if status.employeeLastName is not empty %}
           {{ status.employeeFirstName ~ ' ' ~ status.employeeLastName }}
         {% endif %}
+        {% if status.apiClientId is not empty %}
+          {{ 'Api Client: %api_client_id%'|trans({'%api_client_id%': status.apiClientId}, 'Admin.Global') }}
```
Comment (kpodemski) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/history.html.twig:42]: Could we get the information about the given id_order_history? What if we would do:

` $this->mutationTracker->addMutationForModule('order_history', (int) $history->id, MutationAction::CREATE);`

I would have to hook into the displayAdminOrderSomething and create my own list of changes to the history, right?

Comment (kpodemski) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/history.html.twig:42]: So, instead of joining on API_CLIENT, you could only get the information about a given Mutation row for a given history record and return the information accordingly. It could be api_client, module, etc.

File: src/PrestaShopBundle/Entity/Mutation.php (hunk 7 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Entity/Mutation.php:56]: Yeah actually Doctrine doesn't allow entities without primary keys For regular associations we would use the couple of IDs as primary keys but the thing is that in this case we can't because the associations may be duplicate (same object same id over different mutations in time) So we need this independent ID, it's mandatory and a good practice to have this primary key especially in case we need to handle the DELETION in the future (we won't be able to delete based on associations as some may need to be removed and some not)

I don't know if it's clear 😅

Comment (kpodemski) [reply src/PrestaShopBundle/Entity/Mutation.php:56]: It is, thanks.

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/history.html.twig (hunk 8 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/history.html.twig:42]: Ah you want to also display if a modifiaction was made by a module
This was not implemented so far, right now the CQRS query has been modified to return the API client relation, it is done explicitly as CQRS query results must be explicit in their structure, it doesb't return the mutation directly it bases its SQL query on the table and reformat it in a domain way

TLDR; if we want to handle this it should be done in a dedicated PR/new feature

Comment (kpodemski) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/history.html.twig:42]: Yeah, this would need to have something like this:

```
{% if status.mutationTrackerDetails is not empty %}
{{ status.mutationTrackerDetails }}
{% endif %}
```

```
 LEFT JOIN `' . _DB_PREFIX_ . 'mutation` m ON m.`mutation_table` = "order_history" AND m.`mutation_row_id` = oh.`id_order_history` 
 ```

And we would need to have something to render information based on `mutation.id` if we are keeping the ID anyway.

Formatter could be inside `GetOrderForViewingHandler`

Comment (kpodemski) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/history.html.twig:42]: Thanks to this, instead of informing about API changes, you're providing the core with the information on the actor who made the mutation. Module, API, employee, doesn't matter.

**Review by kpodemski — APPROVED** [review]
I'm approving, but I still have some doubts I shared in this comment: https://github.com/PrestaShop/PrestaShop/pull/35292/files#r1492459582. 

Hopefully we can back to this subject in the future PR 😎

**Review by Hlavtox — APPROVED** [review]

**Review by tleon — APPROVED** [review]

File: src/Core/Domain/Employee/ValueObject/EmployeeId.php [hunk 9]
```diff
@@ -31,7 +31,7 @@
 /**
  * Defines Employee ID with it's constraints.
  */
-class EmployeeId
+class EmployeeId implements EmployeeIdInterface
```
Comment (matks) [inline src/Core/Domain/Employee/ValueObject/EmployeeId.php:34]: BC Break to document

File: src/Core/Mutation/MutationTracker.php [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+use Doctrine\ORM\EntityManagerInterface;
+use PrestaShop\PrestaShop\Core\Context\ApiClientContext;
+use PrestaShop\PrestaShop\Core\Context\EmployeeContext;
+use PrestaShopBundle\Entity\Mutation;
+use PrestaShopBundle\Entity\MutationAction;
+use PrestaShopBundle\Entity\MutatorType;
+
+/**
+ * The mutation track service helps to add mutation from any service in the code base, it automatically
```
Comment (matks) [inline src/Core/Mutation/MutationTracker.php:39]: I suggest to give some examples of mutations that can happen

File: tests/Integration/Behaviour/Features/Context/EmployeeFeatureContext.php [hunk 11]
```diff
@@ -39,8 +41,48 @@ class EmployeeFeatureContext extends AbstractPrestaShopFeatureContext
...
+            defaultShopId: (int) $legacyEmployee->getDefaultShopID(),
+            associatedShopIds: $legacyEmployee->getAssociatedShopIds(),
+            associatedShopGroupIds: $legacyEmployee->getAssociatedShopGroupIds()
+        ));
+    }
+
+    /**
+     * @Given I am not logged in as an employee
+     */
+    public function logsOutBackOffice()
```
Comment (matks) [inline tests/Integration/Behaviour/Features/Context/EmployeeFeatureContext.php:70]: Will be useful 👍

File: tests/Unit/PrestaShopBundle/EventListener/Context/API/ApiClientContextListenerTest.php [hunk 12]
```diff
@@ -71,11 +72,13 @@ public function testBuildBasedOnSecurityToken(string $clientId, array $scopes):
     public function getExpectedClients(): iterable
     {
         yield 'client with scopes' => [
+            42,
```
Comment (matks) [inline tests/Unit/PrestaShopBundle/EventListener/Context/API/ApiClientContextListenerTest.php:75]: You used constants above but not here

**Review by matks — APPROVED** [review]
Good for me, some noblocking comments

File: src/Core/Mutation/MutationTracker.php (hunk 10 above)
Comment (jolelievre) [reply src/Core/Mutation/MutationTracker.php:39]: There are additional details on the `Mutation` entity, combined with the current use case do you think it's still not enough?

File: tests/Unit/PrestaShopBundle/EventListener/Context/API/ApiClientContextListenerTest.php (hunk 12 above)
Comment (jolelievre) [reply tests/Unit/PrestaShopBundle/EventListener/Context/API/ApiClientContextListenerTest.php:75]: Yeah mainly because it's in a provider methods 🤔

File: src/Core/Domain/Employee/ValueObject/EmployeeId.php (hunk 9 above)
Comment (jolelievre) [reply src/Core/Domain/Employee/ValueObject/EmployeeId.php:34]: I added the details in the PR description
